homeassistant/components/xiaomi_miio/alarm_control_panel.py

```diff
@@ -0,0 +1,150 @@
+    entities = []
+    gateway = hass.data[DOMAIN][config_entry.entry_id]
+    entity = XiaomiGatewayAlarm(
+        gateway,
+        f"{config_entry.title} Alarm",
+        config_entry.data.get("model"),
+        config_entry.data.get("mac"),
+        config_entry.data.get("gateway_id"),
+    )
+    entities.append(entity)
+    async_add_entities(entities)
+
+
+class XiaomiGatewayAlarm(AlarmControlPanel):
+    """Representation of the XiaomiGatewayAlarm."""
+
+    def __init__(
+        self, gateway_device, gateway_name, model, mac_address, gateway_device_id
+    ):
+        """Initialize the entity."""
+        self._gateway = gateway_device
+        self._name = gateway_name
+        self._gateway_device_id = gateway_device_id
+        self._unique_id = f"{model}-{mac_address}"
+        self._icon = "mdi:shield-home"
+        self._available = None
+        self._state = None
+
+    @property
+    def unique_id(self):
+        """Return an unique ID."""
+        return self._unique_id
+
+    @property
+    def device_id(self):
+        """Return the device id of the gateway."""
+        return self._gateway_device_id
+
+    @property
+    def device_info(self):
+        """Return the device info of the gateway."""
+        return {
+            "identifiers": {(DOMAIN, self._gateway_device_id)},
+        }
+
+    @property
+    def name(self):
+        """Return the name of this entity, if any."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon to use for device if any."""
+        return self._icon
+
+    @property
+    def available(self):
+        """Return true when state is known."""
+        return self._available
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def supported_features(self) -> int:
+        """Return the list of supported features."""
+        return SUPPORT_ALARM_ARM_AWAY
+
+    async def _try_command(self, mask_error, func, *args, **kwargs):
+        """Call a device command handling error messages."""
+        try:
+            result = await self.hass.async_add_executor_job(
+                partial(func, *args, **kwargs)
+            )
+            _LOGGER.debug("Response received from miio device: %s", result)
+        except DeviceException as exc:
+            _LOGGER.error(mask_error, exc)
+
+    async def async_alarm_arm_away(self, code=None):
+        """Turn on."""
+        await self._try_command(
+            "Turning the alarm on failed: %s", self._gateway.alarm.on
+        )
+
+    async def async_alarm_disarm(self, code=None):
+        """Turn off."""
+        await self._try_command(
+            "Turning the alarm off failed: %s", self._gateway.alarm.off
+        )
+
+    async def async_update(self):
+        """Fetch state from the device."""
+        try:
+            state = await self.hass.async_add_job(self._gateway.alarm.status)
```

Comment:
Replace all `hass.async_add_job` with either `hass.async_create_task` or `hass.async_add_executor_job` depending on the target. If it's a function that should run in the executor thread pool, use `hass.async_add_executor_job`.

Reply:
The gateway here is the wrapped gateway from the gateway module, right? We need to adjust the attribute access, right?

Reply:
I replaced all hass.async_add_job.
The gateway here is **not** the wrapped gateway from the gateway.py file in HomeAssistant.
The gateway here is the gateway class of the python miio module (v0.5.0.1)

Reply:
So I do not think we need to adjust the attribute access